internal/remediation/override.go

```diff
@@ -0,0 +1,279 @@
+		// TODO: CalculateDiff maybe should not be reconstructing patches.
+		// Refactor CalculateDiff, Relaxer, Override to make patches in a more sane way.
+		diff.Deps = make([]manifest.DependencyPatch, len(res.patches))
+		for i, p := range res.patches {
+			diff.Deps[i] = manifest.DependencyPatch{
+				Pkg:          p.PackageKey,
+				Type:         dep.Type{},
+				OrigRequire:  "", // Using empty original to signal this is an override patch
+				OrigResolved: p.OrigVersion,
+				NewRequire:   p.NewVersion,
+				NewResolved:  p.NewVersion,
+			}
+		}
+
+		allResults = append(allResults, diff)
+
+		// If there are any new vulns, try override them as well
+		var newlyAdded []string
+		for _, v := range diff.AddedVulns {
+			if !slices.Contains(res.vulnIDs, v.Vulnerability.ID) {
+				newlyAdded = append(newlyAdded, v.Vulnerability.ID)
+			}
+		}
+
+		if len(newlyAdded) > 0 {
+			go doOverride(append(res.vulnIDs, newlyAdded...)) // No need to clone res.vulnIDs here
+			toProcess++
+		}
+	}
+
+	// Sort and remove duplicate patches
+	slices.SortFunc(allResults, func(a, b resolution.ResolutionDiff) int { return a.Compare(b) })
+	allResults = slices.CompactFunc(allResults, func(a, b resolution.ResolutionDiff) bool { return a.Compare(b) == 0 })
+
+	return allResults, nil
+}
+
+var errOverrideImpossible = errors.New("cannot fix vulns by overrides")
+
+func overridePatchVulns(ctx context.Context, cl client.ResolutionClient, result *resolution.ResolutionResult, vulnIDs []string, opts RemediationOptions) (*resolution.ResolutionResult, []overridePatch, error) {
+	// Try to fix as many vulns in vulnIDs as possible.
+	// returns errOverrideImpossible if 0 vulns are patchable.
+	var effectivePatches []overridePatch
+	for {
+		// Find the relevant vulns affecting each version key.
+		vkVulns := make(map[resolve.VersionKey][]*resolution.ResolutionVuln)
+		for i, v := range result.Vulns {
+			if !slices.Contains(vulnIDs, v.Vulnerability.ID) {
+				continue
+			}
+			seenVks := make(map[resolve.VersionKey]struct{})
+			for _, c := range v.ProblemChains {
+				vk, _ := c.End()
+				if _, seen := seenVks[vk]; !seen {
+					vkVulns[vk] = append(vkVulns[vk], &result.Vulns[i])
+					seenVks[vk] = struct{}{}
+				}
+			}
+			for _, c := range v.NonProblemChains {
+				vk, _ := c.End()
+				if _, seen := seenVks[vk]; !seen {
+					vkVulns[vk] = append(vkVulns[vk], &result.Vulns[i])
+					seenVks[vk] = struct{}{}
+				}
+			}
+		}
+
+		if len(vkVulns) == 0 {
+			// All vulns have been fixed.
+			break
+		}
+
+		newPatches := make([]overridePatch, 0, len(vkVulns))
+
+		for vk, vulnerabilities := range vkVulns {
+			// Consider vulns affecting packages we don't want to change unfixable
+			if slices.Contains(opts.AvoidPkgs, vk.Name) {
+				continue
+			}
+
+			sys := vk.Semver()
+			// Get & sort all the valid versions of this package
+			// TODO: (Maven) skip unlisted versions and versions on other registries
+			versions, err := cl.Versions(ctx, vk.PackageKey)
```

Comment:
Can we break up this function into smaller functions? e.g. getSortedPackageVersions()

Having a lot of this logic inline makes the logic a bit hard to follow.

Reply:
Split `getVersionsGreater()` out, still considering other parts to split out.